### PR TITLE
fix(server): allow initiator-only threads

### LINKED ISSUE #20: CreateThread: accept participants (UUID or @nickname) and resolve at creation time
## Background
Current `CreateThreadRequest` only accepts `participant_ids` (UUID strings). Clients that want to add `@nicknames` must currently do a 2-step flow (`CreateThread` then `AddParticipant`).

This is causing breakage with agyn-cli flows where all `--add` values are nicknames.

## Request
Update **Threads CreateThread** semantics so clients can send *all* initial participants in one call.

## Specification
In `CreateThread`:

### Participants input
- Support a new request field `participants` (repeated `ParticipantIdentifier`) alongside legacy `participant_ids`.
- `participant_ids` and `participants` are **mutually exclusive**; reject if both set.
- Each `participants[]` entry can be either:
  - `participant_id` (UUID), or
  - `participant_nickname` (`@nickname`) resolved server-side.

### Organization context (`organization_id`)
- `organization_id` represents the organization context for thread creation / scoping (not merely nickname-resolution convenience).
- If `organization_id` is omitted, the server should derive it from caller identity where possible (e.g. agent identity via Agents service). If it cannot be derived, the request is rejected (users must provide it).
- The resolved `organization_id` is also used as the scope for resolving any `participant_nickname` values.

### Initiator semantics
- Continue to auto-add the request initiator (from identity metadata) as the first participant by default:
  - If initiator is an agent ⇒ store with `passive=true`
  - Otherwise ⇒ `passive=false`
- All non-initiator initial participants must be stored with `passive=false`.

### Validation
- Duplicate rules:
  - Reject duplicate participants in the input list.
  - If the initiator identity is present in the initial participants list (directly or via nickname resolution), return `InvalidArgument`.
- Allow an empty participant list when initiator metadata is present (thread with just initiator).
- If both participant lists are empty and initiator metadata is missing ⇒ `InvalidArgument`.

## Tests
Add/extend server tests to cover:
- CreateThread with mixed UUID + nickname identifiers.
- Organization_id requirement/derivation behavior.
- Mutual exclusivity (`participant_ids` + `participants`).
- Empty participants + initiator metadata works; empty + no metadata errors.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- allow CreateThread to accept empty participant_ids when initiator metadata is present
- add tests for initiator-only thread creation and missing metadata rejection

## Testing
- buf generate buf.build/agynio/api --path agynio/api/threads/v1 --path agynio/api/notifications/v1 --path agynio/api/identity/v1 --path agynio/api/metering/v1 --path agynio/api/agents/v1
- go vet ./...
- go test ./...
- go build ./...

Fixes #20

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Summary
- allow initiator-only CreateThread when metadata is present
- add tests for empty participant_ids with initiator and missing metadata rejection

## Test & Lint Summary
Commands:
- buf generate buf.build/agynio/api --path agynio/api/threads/v1 --path agynio/api/notifications/v1 --path agynio/api/identity/v1 --path agynio/api/metering/v1 --path agynio/api/agents/v1
- go vet ./...
- go test ./...
- go build ./...

Results:
- Tests: passed (go test ./..., 0 failed, 0 skipped)
- Lint: no errors (go vet ./...)

**Review by noa-lucent — DISMISSED** [review]
Clean, minimal fix. The `initiatorInfoFromContext` call is moved before the empty-participants gate, and the condition is correctly widened to `len(ids) == 0 && !hasInitiator`. No behavioral changes beyond what the issue requires.

Both new tests match the issue specification:
- empty participants + agent initiator → 1 passive participant (happy path)
- empty participants + no metadata → `InvalidArgument` (rejection path)

Test patterns are consistent with the existing suite. LGTM.

**Comment by rowan-stein** [comment]
Scope update (per latest requirements): this PR/issue is being expanded from the empty `participant_ids` guard fix to implementing `CreateThreadRequest.participants` (ParticipantIdentifier: UUID or `@nickname`) + server-side nickname resolution at thread creation time, with `participant_ids` kept as deprecated/legacy.

See updated issue: https://github.com/agynio/threads/issues/20
Upstream proto work tracked in agynio/api#118.

**Comment by casey-brooks** [comment]
## Update
- added CreateThread support for ParticipantIdentifier inputs with mutual exclusivity checks
- resolved nicknames using organization_id (derived for agents when omitted)
- added CreateThread tests for nickname resolution, org derivation, and missing org errors

## Test & Lint Summary
Commands:
- buf generate buf.build/agynio/api --path agynio/api/threads/v1 --path agynio/api/notifications/v1 --path agynio/api/identity/v1 --path agynio/api/metering/v1 --path agynio/api/agents/v1
- go vet ./...
- go test ./...
- go build ./...

Results:
- Tests: passed (go test ./..., 0 failed, 0 skipped)
- Lint: no errors (go vet ./...)

File: internal/server/server.go [hunk 1]
```diff
@@ -70,44 +70,92 @@ func New(store threadStore, notifier notifierPublisher, identity identityResolve
...
+				return nil, status.Errorf(codes.InvalidArgument, "participant_ids[%d]: %v", i, err)
+			}
+			if hasInitiator && id == initiator.ID {
+				return nil, status.Error(codes.InvalidArgument, "participant_ids must not include initiator")
+			}
+			if _, ok := seen[id]; ok {
+				return nil, status.Errorf(codes.InvalidArgument, "participant_ids[%d]: duplicate participant", i)
+			}
+			seen[id] = struct{}{}
+			resolved = append(resolved, store.ParticipantInput{ID: id, Passive: false})
```
Comment (noa-lucent) [inline internal/server/server.go]: [major] **Duplicate validate-and-collect logic.** The post-resolution pattern — initiator check, duplicate check, `seen` insert, `resolved` append — is repeated identically three times: here in the `participant_ids` loop (lines 93–100), in the `ParticipantId` switch case (lines 116–123), and in the `ParticipantNickname` case (lines 137–144).

Extract a small helper that takes the resolved `id`, the loop index, and a field-name prefix, and performs the shared validation + collection:

```go
func (s *Server) addResolvedParticipant(
    id uuid.UUID,
    initiator initiatorInfo,
    hasInitiator bool,
    seen map[uuid.UUID]struct{},
    resolved *[]store.ParticipantInput,
    fieldName string,
    index int,
) error {
    if hasInitiator && id == initiator.ID {
        return status.Errorf(codes.InvalidArgument, "%s must not include initiator", fieldName)
    }
    if _, ok := seen[id]; ok {
        return status.Errorf(codes.InvalidArgument, "%s[%d]: duplicate participant", fieldName, index)
    }
    seen[id] = struct{}{}
    *resolved = append(*resolved, store.ParticipantInput{ID: id, Passive: false})
    return nil
}
```

This removes the triple repetition, centralizes the validation rules, and prevents drift if the policy changes later.

File: internal/server/server_test.go [hunk 2]
```diff
@@ -268,6 +324,203 @@ func TestCreateThreadMissingIdentityMetadataDefaultsActive(t *testing.T) {
...
+		t.Fatalf("expected InvalidArgument, got %s: %s", st.Code(), st.Message())
+	}
+	if st.Message() != "organization_id must be provided for participant_nickname" {
+		t.Fatalf("expected organization_id error, got %s", st.Message())
+	}
+}
+
 func TestCreateThreadRejectsInitiatorInParticipants(t *testing.T) {
 	initiatorID := uuid.New()
 	participantID := uuid.New()
```
Comment (noa-lucent) [inline internal/server/server_test.go:619]: [major] **Missing tests required by the issue spec.** Issue #20 explicitly lists these test scenarios:

1. **Mutual exclusivity** — `participant_ids` and `participants` both set → `InvalidArgument`. There is no test for this, despite the validation at line 80 in `server.go`.

2. **Mixed UUID + nickname identifiers** — a `participants` list containing both `ParticipantId` and `ParticipantNickname` entries in a single call. This is a core new capability and has no coverage.

Please add tests for both.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Two major items to address:

1. **Duplicated validate-and-collect logic** — The initiator check, duplicate check, and `resolved` append pattern is repeated three times across the `participant_ids` loop and both `participants` switch cases. Extract a shared helper to centralize this policy and prevent drift.

2. **Missing required tests** — The issue spec explicitly requires tests for mutual exclusivity (`participant_ids` + `participants` both set) and mixed UUID + nickname identifiers. Neither exists.

The rest of the implementation looks correct: mutual exclusivity gate, lazy org-id resolution, `resolveNickname` extraction, and the `organizationIDForNickname` signature change are all clean. Tests for empty-participants, org-id derivation from agent identity, and user org-id requirement are solid.

File: internal/server/server.go (hunk 1 above)
Comment (casey-brooks) [reply internal/server/server.go]: Extracted shared addResolvedParticipant helper to centralize initiator/duplicate checks and reuse it across participant_ids and participants handling.

**Comment by casey-brooks** [comment]
## Update
- deduplicated CreateThread participant validation with a shared helper
- added CreateThread tests for mutual exclusivity and mixed UUID/nickname participants

## Test & Lint Summary
Commands:
- go vet ./...
- go test ./...
- go build ./...

Results:
- Tests: passed (go test ./..., 0 failed, 0 skipped)
- Lint: no errors (go vet ./...)

**Review by noa-lucent — DISMISSED** [review]
Both requested changes verified in the actual code:

1. **Helper extraction** — `addResolvedParticipant` centralizes initiator check, duplicate check, seen-map insert, and resolved-slice append. All three former call sites now delegate to it. Clean, no-receiver package function — correct since it doesn't need `*Server`.

2. **Missing tests added** — `TestCreateThreadRejectsMixedParticipantFields` covers mutual exclusivity with exact error message assertion. `TestCreateThreadMixedParticipantIdentifiers` covers mixed UUID + nickname in a single `participants` list, verifying participant order, store invocation, and identity resolution.

No new issues. LGTM.

**Review by noa-lucent — APPROVED** [review]
Re-approving after base retarget from `noa/issue-1` to `main`. The diff is identical — same commit (ede49e6), same two files, same code. All prior feedback threads are resolved. No new issues.
